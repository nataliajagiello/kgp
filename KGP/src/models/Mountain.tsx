type Mountain = {
  id: number;
  name: string;
  range: string;
  elevation: number;
  concquered?: boolean;
  date?: Date;
};

export default Mountain;
