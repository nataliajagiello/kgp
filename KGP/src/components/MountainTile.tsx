import React, {useState} from 'react';
import {Text, View, StyleSheet} from 'react-native';
import CheckBox from '@react-native-community/checkbox';
import Mountain from '../models/Mountain';
import AsyncStorage from '@react-native-async-storage/async-storage';
import DateTimePicker, {
  DateTimePickerEvent,
} from '@react-native-community/datetimepicker';
import {formatDate} from '../helpers/dateHelpers';

const styles = StyleSheet.create({
  tile: {
    borderWidth: 1,
    flexDirection: 'row',
  },
  text: {
    fontSize: 20,
    textAlign: 'center',
    fontWeight: 'bold',
    margin: 5,
  },
  mainText: {
    fontSize: 25,
    textAlign: 'center',
    fontWeight: 'bold',
    margin: 5,
  },
  checkboxContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    flex: 1,
  },
  checkbox: {
    transform: [{scaleX: 1.3}, {scaleY: 1.3}],
    marginLeft: 25,
  },
  dataContainer: {flex: 10},
});

const MountainTile = ({
  mountain,
  updateMountain,
}: {
  mountain: Mountain;
  updateMountain: Function;
}) => {
  const [isSelected, setSelection] = useState(mountain.concquered);
  const [date, setDate] = useState<Date>(new Date());
  const [showDatepicker, setShowDatepicker] = useState(false);

  const onChange = (
    event: DateTimePickerEvent,
    selectedDate: Date | undefined,
  ) => {
    setShowDatepicker(false);
    selectedDate && setDate(selectedDate);
    updateData(true, event.type === 'dismissed' ? undefined : selectedDate);
  };

  const handleCheckboxChange = async (value: boolean) => {
    setSelection(value);
    if (value) {
      setShowDatepicker(true);
    } else {
      updateData(false, undefined);
    }
  };

  const updateData = async (
    concqured: boolean,
    concqueredDate: Date | undefined,
  ) => {
    await AsyncStorage.setItem(`${mountain.id}`, `${concqured}`);
    await AsyncStorage.setItem(`date-${mountain.id}`, `${concqueredDate}`);
    mountain.concquered = concqured;
    mountain.date = concqueredDate;
    updateMountain(mountain);
  };

  return (
    <View style={styles.tile}>
      {showDatepicker && (
        <DateTimePicker
          testID="datePicker"
          value={date}
          mode="date"
          is24Hour={false}
          onChange={onChange}
          maximumDate={new Date()}
        />
      )}
      <View style={styles.checkboxContainer}>
        <CheckBox
          testID={`m-${mountain.id.toString()}`}
          value={isSelected}
          onValueChange={(value: boolean) => {
            handleCheckboxChange(value);
          }}
          style={styles.checkbox}
        />
      </View>
      <View style={styles.dataContainer}>
        {mountain?.concquered && mountain?.date && (
          <Text testID="m-date" style={styles.text}>
            {formatDate(mountain.date)}
          </Text>
        )}
        <Text style={styles.text}>{mountain.range}</Text>
        <Text style={styles.mainText}>{mountain.name}</Text>
        <Text style={styles.text}>{mountain.elevation} m n.p.m.</Text>
      </View>
    </View>
  );
};

export default MountainTile;
