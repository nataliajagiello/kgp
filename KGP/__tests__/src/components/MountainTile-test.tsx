import 'react-native';
import React from 'react';

import renderer from 'react-test-renderer';
import {fireEvent, render, act} from '@testing-library/react-native';

import MountainTile from '../../../src/components/MountainTile';
import Mountain from '../../../src/models/Mountain';
//import {DateTimePickerEvent} from '@react-native-community/datetimepicker';

const mountain: Mountain = {
  id: 1,
  elevation: 1200,
  name: 'Super Górka',
  range: 'Pasmo',
};
const updateMountain = jest.fn();

it('renders correctly', async () => {
  const tree = renderer
    .create(
      <MountainTile mountain={mountain} updateMountain={updateMountain} />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders mountain data', async () => {
  const {getByText} = render(
    <MountainTile mountain={mountain} updateMountain={updateMountain} />,
  );
  expect(getByText('Pasmo')).toBeTruthy();
  expect(getByText('Super Górka')).toBeTruthy();
  expect(getByText('1200 m n.p.m.')).toBeTruthy();
});

test('on checkbox click it shows calendar', async () => {
  const {findByTestId} = render(
    <MountainTile mountain={mountain} updateMountain={updateMountain} />,
  );
  const checkbox = await findByTestId('m-1');
  await act(async () => {
    fireEvent(checkbox, 'onValueChange', true);
  });
  const datepicker = await findByTestId('datePicker');
  expect(datepicker).toBeTruthy();
});
