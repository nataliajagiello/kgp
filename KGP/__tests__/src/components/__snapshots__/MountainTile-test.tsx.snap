// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "borderWidth": 1,
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityRole="checkbox"
      accessibilityState={
        Object {
          "checked": false,
          "disabled": false,
        }
      }
      accessibilityValue={
        Object {
          "text": "off",
        }
      }
      accessible={true}
      pointerEvents="auto"
    >
      <RNCCheckbox
        forwardedRef={null}
        onValueChange={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "height": 32,
              "width": 32,
            },
            undefined,
          ]
        }
        value={false}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
          "margin": 5,
          "textAlign": "center",
        }
      }
    >
      Pasmo
    </Text>
    <Text
      style={
        Object {
          "fontSize": 25,
          "fontWeight": "bold",
          "margin": 5,
          "textAlign": "center",
        }
      }
    >
      Super GÃ³rka
    </Text>
    <Text
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
          "margin": 5,
          "textAlign": "center",
        }
      }
    >
      1200
       m n.p.m.
    </Text>
  </View>
</View>
`;
